Implement gen_send_cfunc and gen_send_iseq

noahgibbs opened the pull request:
YJIT-side codegen methods aren't implemented yet - we basically just return None for now.

Keyword-based iseq calls aren't fully ported. A lot of the C code is still there, commented out, next to a todo!().

noahgibbs committed:
Implement gen_send_cfunc and gen_send_iseq

maximecb commented on yjit.c:
```diff
@@ -304,6 +310,42 @@ rb_get_cme_def_body_optimized_type(rb_callable_method_entry_t* cme)
+}
+
+int
+rb_get_mct_argc(rb_method_cfunc_t *mct)
+{
+    return mct->argc;
+}
+
+void*
+rb_get_mct_func(rb_method_cfunc_t *mct)
+{
+    return (void*)mct->func; // this field is defined as type VALUE (*func)(ANYARGS)
+}
+
+int
+rb_get_builtin_argc(struct rb_builtin_function* bi)
+{
+    return bi->argc;
+}
+
+void*
+rb_get_builtin_func_ptr(struct rb_builtin_function* bi)
+{
+    return bi->func_ptr;
+}
```
@XrXr maybe we should export some of these structs because it's going to become hard to keep track of what these getters do.

maximecb replied:
This is going to make for unreadable code :/ Not your fault Noah, just saying not exporting the C struct defs is going to make the Rust code hard to read.

noahgibbs replied:
I agree. This is what I was worried about when we talked earlier about having the little C getters - there aren't as many as I was afraid of, but there are a *lot* and they're pretty hard to keep track of. Exporting the C struct defs should be okay in terms of future changes as long as we regenerate bindings for them frequently, I think.

XrXr replied:
Yeah, for this one in particular it's probably good to export it since the definition is simple and there are no ifdefs in the body of the struct.

maximecb commented on yjit.c:
```diff
@@ -314,16 +356,62 @@ rb_get_iseq_body_iseq_encoded(rb_iseq_t* iseq) {
+
+// Returns whether the iseq only needs positional (lead) argument setup.
+bool
+rb_iseq_lead_only_arg_setup_p(const rb_iseq_t *iseq)
```
Could have a more descriptive name maybe, `rb_iseq_needs_lead_args_only` or something like that?

maximecb commented on yjit/bindgen/src/main.rs:
```diff
@@ -64,6 +64,8 @@ fn main() {
         .allowlist_type("vm_call_flag_bits") // So instead we include the other enum and do the bit-shift ourselves
 
+        .allowlist_var("VM_BLOCK_HANDLER_NONE")
+
```
We should probably get in the habit of commenting on which files these definitions come from

noahgibbs replied:
I'll go through and add them. Huh. And it looks like we're currently parsing yjit.c. So we *could* move a lot of the cruby.rs extern definitions into bindgen, but only if we're okay with keeping the rb_ prefix.

maximecb replied:
I think we're ok with the `rb_` prefix.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3311,19 +3309,20 @@ fn jit_protected_callee_ancestry_guard(jit: &mut JITState, cb: &mut CodeBlock, c
 // known_recv_klass is non-NULL when the caller has used jit_guard_known_klass().
 // See yjit_reg_method().
-typedef bool (*method_codegen_t)(jitstate_t *jit, ctx_t *ctx, const struct rb_callinfo *ci, const rb_callable_method_entry_t *cme, rb_iseq_t *block, const int32_t argc, VALUE *known_recv_klass);
+type CCodeGenMethod = fn(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &mut OutlinedCb, ci: *const rb_callinfo, cme: *const rb_callable_method_entry_t, block: Option<IseqPtr>, argc: i32, known_recv_class: *const VALUE) -> bool;
```
Maybe rename to `MethodGenFn`, it's generating code for a specific method, not specifically C code.

noahgibbs replied:
Right now we have a CodeGenFn for generating assembly for YARV opcodes. I feel like CodeGenFn vs MethodGenFn is pretty murky. Maybe I could make it a MethodGenFn and rename CodeGenFn to something like OpcodeGenFn?

maximecb replied:
Let's go with `InsnGenFn`. Opcode means just the constant specifying which operation to perform, whereas instruction refers to opcode + operands.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3497,119 +3497,117 @@ fn c_method_tracing_currently_enabled(jit: &JITState) -> bool
-            }
+    let cme_def = unsafe { get_cme_def(cme) };
+    let codegen_p = lookup_cfunc_codegen(cme_def);
+    if ! matches!(codegen_p, None) {
```
I think you should use `codegen_p.is_some()` here

maximecb commented on yjit/src/codegen.rs:
Here could also use `block.is_some()`, or the `if let` form: https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html

noahgibbs replied:
Ah, I hadn't known "if let". That's really useful here, thanks!

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3621,76 +3619,81 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
 
     // Pop the C function arguments from the stack (in the caller)
-    ctx.stack_pop(argc + 1);
+    ctx.stack_pop((argc + 1).try_into().unwrap());
```
Can just use `.into()` here.

noahgibbs replied:
I do in some places. But for that one it gives a "From&lt;i32&gt; not implemented for usize" error, alas.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3621,76 +3619,81 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
-        {
-            let stack_opnd = mem_opnd(64, RAX, -(argc + 1 - i) * SIZEOF_VALUE);
+        for i in 0..=argc as usize { // "as usize?" Yeah, you can't index an array by an i32.
+            let stack_opnd = mem_opnd(64, RAX, -(argc + 1 - (i as i32)) * sizeof_value_i32);
```
If `sizeof_value_i32` is a constant, it should be in all caps

noahgibbs replied:
I wasn't sure about etiquette for that. Anything without a "let mut" won't change, so in theory they're nearly all constants. But I don't *think* we want them all in all-caps.

maximecb replied:
we should define a SIZEOF_VALUE_I32 constant in `cruby.rs`

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3621,76 +3619,81 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
-        for (int32_t i = 0; i < argc + 1; ++i)
-        {
-            let stack_opnd = mem_opnd(64, RAX, -(argc + 1 - i) * SIZEOF_VALUE);
+        for i in 0..=argc as usize { // "as usize?" Yeah, you can't index an array by an i32.
```
The loop originally went to `< argc + 1`, I think this one only goes up to `< argc`.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3700,116 +3703,102 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
+    match shape {
+        BranchShape::Next0 | BranchShape::Next1 => unreachable!(),
+        BranchShape::Default => {
+            mov(cb, REG0, const_ptr_opnd(target0.raw_ptr()));
```
There's a `code_ptr_opnd` so we can avoid the .raw_ptr() here.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3700,116 +3703,102 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
+            first_opcode == (OP_OPT_INVOKEBUILTIN_DELEGATE_LEAVE as i32) &&
+            second_opcode == (OP_LEAVE as i32) && bip;
+    };
  }
```
I think we'll want to keep this bit of code in C, as you did with the other predicate above.

noahgibbs replied:
Fair. These are fairly messy in Rust. I'll switch them over.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3700,116 +3703,102 @@ fn gen_send_cfunc(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
+    }
+    unsafe {
+        let iseq_encoded = get_iseq_body_iseq_encoded(iseq);
+        let second_value = *(iseq_encoded.offset(SIZEOF_VALUE as isize));
```
Probably also want to keep this in C

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -3924,50 +3922,62 @@ fn gen_send_iseq(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
+                mov(cb, C_ARG_REGS[0], REG_EC);
+
+                // Copy self and arguments
+                for i in 0..=builtin_argc {
```
I think this is also missing the +1 from the above loop end condition

noahgibbs replied:
The two you comment on both have a "..=" instead of a "..". That's a Rust inclusive range, so it should (I believe) also hit the top element. https://doc.rust-lang.org/nightly/std/ops/struct.RangeInclusive.html

maximecb replied:
Did not know about the inclusive range 👍

noahgibbs commented:
I've addressed most of the feedback or let you know why I think the current way is okay (can see if you agree :-) ). I'll talk to you when you're around about the all-caps constant thing and the MethodGenFn final names. The bindgen/src/main.rs reorg is maybe a little extreme, but it seems good to just fix it now - LMK if this way works for you.

noahgibbs commented:
Okay, fixed the all-caps constant and function-type-rename parts.

noahgibbs commented:
For now I have *not* exported any of the C structures through bindgen. I figure there's probably already enough going on in this PR and that one should be separate. LMK if you'd prefer it done immediately in this PR instead.

noahgibbs commented:
Rebasing onto latest.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -4076,64 +4041,73 @@ fn gen_send_iseq(jit: &JITState, ctx: &Context, ci: * const rb_callinfo, cme: *
+            mov(cb, mem_opnd(64, REG_CFP, RUBY_OFFSET_CFP_BLOCK_CODE), REG0);
+        },
+        _ => (),
     }
```
This is the perfect place for the `if let` form

noahgibbs replied:
Fixed.

maximecb commented on yjit/bindgen/src/main.rs:
```diff
@@ -61,72 +72,95 @@ fn main() {
+        .allowlist_function("rb_iseq_(get|set)_yjit_payload")
         .allowlist_function("rb_iseq_pc_at_idx")
         .allowlist_function("rb_iseq_opcode_at_pc")
+        .allowlist_function("rb_yjit_mark_writable")
+        .allowlist_function("rb_yjit_mark_executable")
+        .allowlist_function("rb_yjit_get_page_size")
```
I think we could allowlist `rb_yjit_*` because those are all our functions.

noahgibbs replied:
Doing that the obvious way causes bindgen to emit a ton of bitfield-access code in Rust (I've included a small excerpt below.) I don't think we can manually blocklist the \_\_Bindgen structs since we're already doing that with all types starting with "\_\_" and it's clearly not blocking them.

I'll see if I can find a way to get it not to do that. I'm worried that we'll have to blocklist every bitfield structure it sees referenced -- I'm pretty sure that's why we blocklist some of the opaque CRuby types.

~~~
#[repr(C)]
#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
impl<Storage> __BindgenBitfieldUnit<Storage> {
    #[inline]
    pub const fn new(storage: Storage) -> Self {
        Self { storage }
    }
}
impl<Storage> __BindgenBitfieldUnit<Storage>
where
    Storage: AsRef<[u8]> + AsMut<[u8]>,
{
    #[inline]
    pub fn get_bit(&self, index: usize) -> bool {
        debug_assert!(index / 8 < self.storage.as_ref().len());
        let byte_index = index / 8;
        // ... lots more code follows ...
~~~

noahgibbs replied:
Yeah, looks like we'd need to blocklist anything it can transitively find. There are only a few rb_yjit functions, so I'm not sure how it picks up so much garbage. But if we just allow those, it still picks up 2,000 lines of extra in cruby_bindings, plus we have to blocklist a lot of things to be able to compile (e.g. the C CodeBlock structure gets picked up, and it conflicts with the Rust one -- same for a bunch of other same-name C structs.)

noahgibbs replied:
Ah, okay. We tell it to parse yjit.c, so it's picking up yjit.h, so it's getting a lot of CRuby and YJIT structures from there. Some of this will be decreased by trimming away C code (which you're already doing, thanks for that.) But it means if we just allowlist_function("rb\_yjit\_\.*") then it's going to grab every type mentioned in an rb\_yjit\_ prototype in yjit.h.

D'oh!

maximecb replied:
Cool cool cool. I'll trim more of the old C code today. We can worry about allowlisting `rb_yjit_` functions later.

maximecb commented:
I just put two really minor comments but everything looks great. I agree that exporting structs should be a separate issue/PR. We might want to discuss it with Alan tomorrow or in the meeting Monday.

noahgibbs commented:
I'll plan on another PR to use bindgen for rb_builtin_function.

maximecb commented:
Good work Noah :)
